File: src/main/java/akostenko/aicars/drawing/TrackSectionImg.java
package akostenko.aicars.drawing;

import static java.lang.Math.PI;
import static java.util.stream.Collectors.toList;

import akostenko.aicars.math.Decart;
import akostenko.aicars.math.Polar;
import akostenko.aicars.track.TrackBorder;
import akostenko.aicars.track.TrackSection;

import org.newdawn.slick.Color;

import java.util.Collection;

public class TrackSectionImg {

    public static Collection<Line> get(TrackSection section, double trackWidth, Scale scale, Color color, Decart camera) {
        Decart cameraPx = camera.negative().multi(scale.getPixels()/scale.getSize());
        if (section.isStraight()) {
            return getStraightBorderLines(section, trackWidth).stream()
                    .map(line -> line.scale(scale))
                    .map(line -> line.position(cameraPx, color, 3))
                    .collect(toList());
        } else {
            throw new IllegalArgumentException("Not implemented");
        }
    }

    private static Collection<LinesBuilder.LocalLine> getStraightBorderLines(TrackSection section, double trackWidth) {
        Polar sectionStart = section.start().toPolar();
        Polar sectionEnd = section.start().toPolar().plus(new Polar(section.length(), section.heading()));

        Polar rightBorder = new Polar(trackWidth / 2, section.heading() + PI / 2);
        Polar leftBorder = new Polar(trackWidth / 2, section.heading() - PI / 2);
        return new LinesBuilder()
                .between(sectionStart.plus(rightBorder), sectionEnd.plus(rightBorder))
                .between(sectionStart.plus(leftBorder), sectionEnd.plus(leftBorder))
                .build();
    }

    public static Collection<TrackBorder> getBorders(TrackSection section, double trackWidth) {
        if (section.isStraight()) {
            return getStraightBorderLines(section, trackWidth).stream()
                    .map(TrackSectionImg::createBorder)
                    .collect(toList());
        } else {
            throw new IllegalArgumentException("Not implemented");
        }
    }

    private static TrackBorder createBorder(LinesBuilder.LocalLine line) {
        return new TrackBorder(line.from(), line.to());
    }

}

File: src/main/kotlin/akostenko/aicars/drawing/TrackSectionImg.kt
package akostenko.aicars.drawing

import akostenko.aicars.math.Decart
import akostenko.aicars.math.Polar
import akostenko.aicars.track.TrackBorder
import akostenko.aicars.track.TrackSection
import org.newdawn.slick.Color
import java.lang.Math.PI

object TrackSectionImg {

    fun build(section: TrackSection, scale: Scale, color: Color, camera: Decart): Collection<Line> {
        val cameraPx = scale.to(-camera).toDecart()
        if (section.isStraight) {
            return getStraightBorderLines(section)
                    .map { line -> line.scale(scale) }
                    .map { line -> line.place(cameraPx, color, 3f) }
        } else {
            return getArcBorderLines(section)
                    .map { arc -> arc.scale(scale) }
                    .map { arc -> arc.place(cameraPx, color, 3f)}
        }
    }

    private fun getStraightBorderLines(section: TrackSection): Collection<LinesBuilder.RelativeLine> {
        val sectionStart = section.start.toPolar()
        val sectionEnd = section.start.toPolar().plus(Polar(section.length, section.heading))

        val rightBorderOffset = Polar(section.width / 2, section.heading + PI/2)
        val leftBorderOffset = Polar(section.width / 2, section.heading - PI/2)
        return LinesBuilder()
                .between(sectionStart + rightBorderOffset, sectionEnd + rightBorderOffset)
                .between(sectionStart+ leftBorderOffset, sectionEnd + leftBorderOffset)
                .build()
    }

    private fun getArcBorderLines(section: TrackSection): Collection<LinesBuilder.RelativeArc> {
        val center = section.start + Polar(section.radius, section.heading + PI/2)
        var from = (section.start - center).toPolar().d - if (section.radius < 0) PI else 0.0
        var to = from + section.length / section.radius
        if (section.radius < 0) {
            val t = to;
            to = from;
            from = t;
        }
        return listOf(LinesBuilder.RelativeArc(center, section.radius - section.width/2, from, to),
                LinesBuilder.RelativeArc(center, section.radius + section.width/2, from, to))
    }

    fun getBorders(section: TrackSection): Collection<TrackBorder> {
        if (section.isStraight) {
            return getStraightBorderLines(section)
                    .map({ createBorder(it) })
        } else {
            throw IllegalArgumentException("Not implemented")
        }
    }

    private fun createBorder(line: LinesBuilder.RelativeLine): TrackBorder {
        return TrackBorder(line.from, line.to)
    }

}
